pkgdb2/lib/__init__.py

```diff
@@ -1529,6 +1645,236 @@ def add_branch(session, clt_from, clt_to, user):
+        _validate_pkg(session, clt_to.version, package.name)
+
+    action = model.AdminAction(
+        package_id=package.id,
+        collection_id=clt_to.id,
+        from_collection_id=clt_from.id,
+        user=user.username,
+        _status=status,
+        action='request.branch',
+    )
+
+    session.add(action)
+
+    pkgdb2.lib.utils.log(
+        session,
+        package=package,
+        topic='package.branch.request',
+        message=dict(
+            agent=user.username,
+            package=package.to_json(acls=False),
+            collection_from=clt_from.to_json(),
+            collection_to=clt_to.to_json(),
+        )
+    )
+
+    msg = 'Branch %s requested for user %s' % (
+        clt_to.branchname, user.username)
+
+    # If user is packager -> checked with _validate_poc()
+    # If clt_to is Fedora
+    # If user has commit/approveacls on pkg_name
+    # Then automatically grant the branch request
+    if clt_to.name == 'Fedora':
+        maintainer = False
+        for pkglist in get_acl_package(session, package.name):
+            for acl in pkglist.acls:
+                if acl.fas_name == user.username \
+                        and acl.acl in ['commit', 'approveacls']:
+                    maintainer = True
+                    break
+            if maintainer:
+                break
+
+        if maintainer:
+            for acl in ['commit', 'watchbugzilla',
+                        'watchcommits', 'approveacls']:
+                set_acl_package(
+                    session,
+                    pkg_name=package.name,
+                    pkg_branch=clt_to.branchname,
+                    pkg_user=user.username,
+                    acl=acl,
+                    status='Approved',
+                    user=user,
+                    force=True,
+                )
+            msg = 'Branch %s created for user %s' % (
+                clt_to.branchname, user.username)
+
+            # The branch is created, so the action has been approved
+            action.status = 'Approved'
+            session.add(action)
+
+    return msg
+
+
+def add_new_package_request(
+        session, pkg_name, pkg_summary, pkg_description, pkg_status,
+        pkg_collection, pkg_poc, user, pkg_review_url,
+        pkg_upstream_url=None, pkg_critpath=False):
+    """ Create a new Package request in the database.
+
+    :arg session: session with which to connect to the database.
+    :arg pkg_name: the name of the package.
+    :arg pkg_summary: a summary description of the package.
+    :arg pkg_description: the description of the package.
+    :arg pkg_status: the status of the package.
+    :arg pkg_collection: the collection in which had the package.
+    :arg pkg_poc: the point of contact for this package in this collection
+    :arg user: the user performing the action
+    :kwarg pkg_review_url: the url of the review-request on the bugzilla
+    :kwarg pkg_upstream_url: the url of the upstream project.
```

Comment:
Would we really need this, as I think we update this from the spec file later anyway?